File: src/main/java/gg/uhc/migration/Entry.java
package gg.uhc.migration;

import gg.uhc.migration.command.MigrationCommand;
import gg.uhc.migration.configuration.*;
import org.bukkit.configuration.InvalidConfigurationException;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.plugin.java.JavaPlugin;

public class Entry extends JavaPlugin {

    @Override
    public void onEnable() {
        FileConfiguration configuration = getConfig();
        configuration.options().copyDefaults(true);
        saveConfig();

        AreaListReader areaListReader = new AreaListReader(new AreaReader(new RangeReader(new RangeParser())));
        PositiveIntegerReader integerReader = new PositiveIntegerReader(false);
        MessageSenderReader senderReader = new MessageSenderReader();
        LongListReader longListReader = new LongListReader();
        PlayerStringsReader stringsReader = new PlayerStringsReader();

        try {
            MigrationRunnableFactory factory = new MigrationRunnableFactory();

            // setup factory with config settings
            factory
                    // areas
                    .areas(areaListReader.readFromSection(configuration, "areas"))

                    // phase timers
                    .damagePhaseTime(integerReader.readFromSection(configuration, "seconds for damaging"))
                    .warningPhaseTime(integerReader.readFromSection(configuration, "seconds for warning"))

                    // senders
                    .damageSender(senderReader.readFromSection(configuration, "damage sender"))
                    .phaseStartSender(senderReader.readFromSection(configuration, "phase change sender"))
                    .updatesSender(senderReader.readFromSection(configuration, "timer update sender"))

                    // damage time
                    .secondsPerHalfHeart(integerReader.readFromSection(configuration, "seconds for half heart"))
                    .notificationTicks(longListReader.readFromSection(configuration, "notification seconds"))

                    // strings
                    .playerStrings(stringsReader.readFromSection(configuration, "messages"));

            getCommand("migration").setExecutor(new MigrationCommand(this, factory));
        } catch (InvalidConfigurationException e) {
            e.printStackTrace();
            setEnabled(false);
        }
    }
}

File: src/main/java/gg/uhc/migration/Entry.kt
package gg.uhc.migration

import gg.uhc.migration.command.MigrationCommand
import gg.uhc.migration.configuration.*
import gg.uhc.migration.selection.AreaSelection
import org.bukkit.configuration.InvalidConfigurationException
import org.bukkit.plugin.java.JavaPlugin

class Entry : JavaPlugin() {
    override fun onEnable() {
        config.options().copyDefaults(true)
        saveConfig()

        try {
            val areas = config.toAreaList("areas")

            val damagePhaseSeconds = config.toInteger("seconds for damaging")
            val warningPhaseSeconds = config.toInteger("seconds for warning")

            val damageSender = config.toMessageSender("damage sender")
            val phaseStartSender = config.toMessageSender("phase change sender")
            val updatesSender = config.toMessageSender("timer update sender")

            val secondsPerHalfHeart = config.toInteger("seconds for half heart")
            val notificationTicks = config.toLongList("notification seconds")

            val messages = config.toCustomMessages("messages")

            val factory = { s: AreaSelection ->
                s.setPotentialAreas(areas)

                MigrationRunnable(
                    selection = s,
                    phaseStartSender = phaseStartSender,
                    updatesSender = updatesSender,
                    damageSender = damageSender,
                    warningPhaseSeconds = warningPhaseSeconds,
                    damagePhaseSeconds = damagePhaseSeconds,
                    secondsForHalfHeart = secondsPerHalfHeart,
                    notificationTicks = notificationTicks,
                    messages = messages
                )
            }

            getCommand("migration").executor = MigrationCommand(this, factory)
        } catch (e: InvalidConfigurationException) {
            e.printStackTrace()
            isEnabled = false
        }
    }
}
